aptos-move/aptos-vm/src/move_vm_ext/warm_vm_cache.rs

```diff
@@ -0,0 +1,162 @@
+        vm_config: VMConfig,
+        resolver: &impl AptosMoveResolver,
+    ) -> VMResult<MoveVM> {
+        WARM_VM_CACHE.get(natives, vm_config, resolver)
+    }
+
+    fn get(
+        &self,
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        vm_config: VMConfig,
+        resolver: &impl AptosMoveResolver,
+    ) -> VMResult<MoveVM> {
+        let _timer = TIMER.timer_with(&["warm_vm_get"]);
+        let id = {
+            let _timer = TIMER.timer_with(&["get_warm_vm_id"]);
+            WarmVmId::new(&vm_config, resolver)?
+        };
+
+        if let Some(vm) = self.cache.read().get(&id) {
+            let _timer = TIMER.timer_with(&["warm_vm_cache_hit"]);
+            return Ok(vm.clone());
+        }
+
+        {
+            let _timer = TIMER.timer_with(&["warm_vm_cache_miss"]);
+            let mut cache_locked = self.cache.write();
+            if let Some(vm) = cache_locked.get(&id) {
+                // Another thread has loaded it
+                return Ok(vm.clone());
+            }
+
+            let vm = MoveVM::new_with_config(natives, vm_config)?;
+            Self::warm_vm_up(&vm, resolver);
+
+            // Not using LruCache because its `::get()` requires &mut self
+            if cache_locked.len() >= WARM_VM_CACHE_SIZE {
+                cache_locked.clear();
+            }
+            cache_locked.insert(id, vm.clone());
+            Ok(vm)
+        }
+    }
+
+    fn warm_vm_up(vm: &MoveVM, resolver: &impl AptosMoveResolver) {
+        let _timer = TIMER.timer_with(&["vm_warm_up"]);
+
+        // Loading `0x1::account` and its transitive dependency into the code cache.
+        //
+        // This should give us a warm VM to avoid the overhead of VM cold start.
+        // Result of this load could be omitted as this is a best effort approach and won't hurt if that fails.
+        //
+        // Loading up `0x1::account` should be sufficient as this is the most common module
+        // used for prologue, epilogue and transfer functionality.
+        let _ = vm.load_module(
+            &ModuleId::new(CORE_CODE_ADDRESS, ident_str!("account").to_owned()),
+            &resolver,
+        );
+    }
+}
+
+#[derive(Eq, Hash, PartialEq)]
+struct WarmVmId {
+    vm_config: Bytes,
+    core_packages_registry: Option<Bytes>,
+}
+
+impl WarmVmId {
+    fn new(vm_config: &VMConfig, resolver: &impl AptosMoveResolver) -> VMResult<Self> {
+        Ok(Self {
+            vm_config: Self::vm_config_bytes(vm_config),
+            core_packages_registry: Self::framework_id_bytes(resolver)?,
+        })
+    }
+
+    fn vm_config_bytes(vm_config: &VMConfig) -> Bytes {
+        let _timer = TIMER.timer_with(&["serialize_vm_config"]);
+        bcs::to_bytes(vm_config)
+            .expect("Failed to serialize VMConfig.")
+            .into()
+    }
+
+    fn framework_id_bytes(resolver: &impl AptosMoveResolver) -> VMResult<Option<Bytes>> {
```

Comment:
probably should call it core_package instead of framework, since it checks everything under 0x1, also not sure if we care about 0x2 and 0x3.

also not sure if there's a way to directly get the hash value from the state key directly